pwnlib/util/misc.py

```diff
@@ -184,6 +185,63 @@ def which(name, all = False, path=None):
     else:
         return None
 
+
+def normalize_argv_env(argv, env, log, level=2):
+    #
+    # Validate argv
+    #
+    # - Must be a list/tuple of strings
+    # - Each string must not contain '\x00'
+    #
+    argv = argv or []
+    if isinstance(argv, (six.text_type, six.binary_type)):
+        argv = [argv]
+
+    if not isinstance(argv, (list, tuple)):
+        log.error('argv must be a list or tuple: %r' % argv)
+
+    if not all(isinstance(arg, (six.text_type, bytes, bytearray)) for arg in argv):
+        log.error("argv must be strings or bytes: %r" % argv)
+
+    # Create a duplicate so we can modify it
+    argv = list(argv)
+
+    for i, oarg in enumerate(argv):
+        arg = packing._need_bytes(oarg, level, 0x80)  # ASCII text is okay
+        if b'\x00' in arg[:-1]:
+            log.error('Inappropriate nulls in argv[%i]: %r' % (i, oarg))
+        argv[i] = bytearray(arg.rstrip(b'\x00'))
+
+    #
+    # Validate environment
+    #
+    # - Must be a dictionary of {string:string}
+    # - No strings may contain '\x00'
+    #
+
+    # Create a duplicate so we can modify it safely
+    env2 = []
+    if hasattr(env, 'items'):
+        env_items = env.items()
+    else:
+        env_items = env
+    if env:
+        for k,v in env_items:
+            if not isinstance(k, (bytes, six.text_type)):
+                log.error('Environment keys must be strings: %r' % k)
+            if not isinstance(k, (bytes, six.text_type)):
+                log.error('Environment values must be strings: %r=%r' % (k,v))
+            k = packing._need_bytes(k, level, 0x80)  # ASCII text is okay
+            v = packing._need_bytes(v, level, 0x80)  # ASCII text is okay
+            if b'\x00' in k[:-1]:
```

Comment:
NUL should not appear anywhere in the key, since the effective format is `KEY=VALUE\x00`